### PR TITLE
WIP - LLVM Bitcode Disassembler for rasm2

### LINKED ISSUE #2778: Add LLVM's BitCode parser and disassembler
```file header is: dec0170b```
See also llvm-dis and `clang -emit-llvm -c foo.c -o foo.bc`

### PR DESCRIPTION
**Pull request by philoinovsky**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

Impletmented the prototype of LLVM Bitcode Disassembler, which reads bitcode stream from stdin and writes .ll stream to stdout

**Test plan**

...

**Closing issues**

closes #2778

### CODE REVIEW

File: libr/asm/p/asm_llvm.c [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+#include <r_types.h>
+#include <r_lib.h>
+#include <r_asm.h>
+
+//-- Primitive Reads ------------------------------------------------
+//Decode a 6-bit encoded character.
+static ut8 getbool(ut8 **bp){
+    ut8 lo = r_read_at_le8(*bp);
+    *bp += 1;
+    return lo & 0x01;
```
Comment (radare) [inline libr/asm/p/asm_llvm.c:15]: use tabs

File: libr/asm/p/asm_llvm.c [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+}
+
+static ut8 char6(ut8 **bp){
+    ut8 lo = r_read_at_le8(*bp);
+    *bp += 6;
+    if(0  <= lo & lo <= 25) return (lo + 97);
```
Comment (radare) [inline libr/asm/p/asm_llvm.c:21]: unnecessary () in return statemnt and wrong spacings before ( and such

File: libr/include/r_asm.h [hunk 3]
```diff
@@ -267,6 +267,7 @@ extern RAsmPlugin r_asm_plugin_xap;
 extern RAsmPlugin r_asm_plugin_xcore_cs;
 extern RAsmPlugin r_asm_plugin_xtensa;
 extern RAsmPlugin r_asm_plugin_z80;
+extern RAsmPlugin r_asm_plugin_llvm;
```
Comment (radare) [inline libr/include/r_asm.h:270]: not sure if we want to name it llvm or bitcode or llil, comments here?

**Comment by radare** [comment]
please rebase and make it green

File: libr/include/r_asm.h (hunk 3 above)
Comment (XVilka) [reply libr/include/r_asm.h:270]: llvm is fine, llil is misleading, bitcode is too vague

**Comment by radare** [comment]
Much wow such wording

**Comment by radare** [comment]
Make it green. All the CI is unhappy

**Comment by XVilka** [comment]
@philoinovsky how is it going with this PR?

**Comment by philoinovsky** [comment]
> @philoinovsky how is it going with this PR?

Final exam is coming, will continue to work on it after exam

**Comment by XVilka** [comment]
Good luck with the exam!

**Comment by ret2libc** [comment]
Also please remember to merge master into your branch before adding other changes, as I see this PR is based on a quite old master, thanks!

**Comment by radare** [comment]
ping

**Comment by MaskRay** [comment]
As a [regular LLVM contributor](https://github.com/llvm/llvm-project/graphs/contributors), I don't think it is feasible to [read LLVM bitcode](https://github.com/llvm/llvm-project/tree/master/llvm/lib/Bitcode/Reader) with 200 lines of code. You can implement the feature as a plugin calling into libLLVMBitReader.

**Comment by XVilka** [comment]
@MaskRay yes, thank you. This PR is unfinished and seems to be abandoned. This is why "draft" status, WIP mark, and the "help-wanted" label.

**Comment by kazarmy** [comment]
Should be a new file format under `libr/bin/format` i think, if this should be in core.

**Comment by ret2libc** [comment]
Given what said by @MaskRay , wouldn't it be easier to move this to radare2-extras and link against llvm libraries? This is of course not possible in radare2 repository, but for extras it is a viable option and it would allow to have a working LLVM disassembler.

**Comment by ret2libc** [comment]
Hi @philoinovsky , do you still plan to work on this or shall we close this PR?

**Comment by trufae** [comment]
I would suggest moving this implementation to extras in case the author (or someone else) wants to contionue working on it.

**Pull request closed without merging**